Source/DafnyCore/AST/IndentationFormatter.cs

```diff
@@ -0,0 +1,2290 @@
+    previousTrivia += token.LeadingTrivia;
+    var lastNL = previousTrivia.LastIndexOf('\n');
+    var lastCR = previousTrivia.LastIndexOf('\r');
+    if (lastNL >= 0 || lastCR >= 0) {
+      // If the leading trivia contains an inline comment after the last newline, it can change the position.
+      var lastCharacterAfterNewline = Math.Max(lastNL, lastCR) + 1;
+      var lastTrivia = previousTrivia.Substring(lastCharacterAfterNewline);
+      if (lastTrivia.IndexOf("*/", StringComparison.Ordinal) >= 0) {
+        return lastTrivia.Length;
+      }
+      if (posToIndentLineBefore.TryGetValue(token.pos, out var indentation)) {
+        return indentation;
+      }
+      if (token.Prev != null &&
+          posToIndentAfter.TryGetValue(token.Prev.pos, out var indentation2)) {
+        return indentation2;
+      }
+      return defaultIndent;
+    }
+    // No newline, so no re-indentation.
+    if (token.Prev != null) {
+      return GetNewTokenVisualIndent(token.Prev, defaultIndent) + token.Prev.val.Length + previousTrivia.Length;
+    }
+
+    return token.col - 1;
+  }
+
+  private static int GetTrailingSpace(IToken token) {
+    var c = 0;
+    while (c < token.TrailingTrivia.Length && token.TrailingTrivia[c] == ' ') {
+      c++;
+    }
+
+    return c;
+  }
+
+  // Given a token such as `var ` immediately followed by another token
+  // returns the indent so that everything after it is aligned with the first token.
+  private int GetRightAlignIndentAfter(IToken token, int indentFallback) {
+    var trailingSpace = GetTrailingSpace(token);
+    return GetNewTokenVisualIndent(token, indentFallback) + token.val.Length + trailingSpace;
+  }
+
+  private static readonly Regex FollowedByNewlineRegex = new Regex("^[ \t]*([\r\n]|//)");
+
+  private static bool IsFollowedByNewline(IToken token) {
+    return FollowedByNewlineRegex.IsMatch(token.TrailingTrivia);
+  }
+
+  #endregion
+
+  #region All SetIndent* methods
+
+  // 'before' is the hypothetical indentation of a comment before that token, and of the previous token if it does not have a set indentation
+  // 'sameLineBefore' is the hypothetical indentation of this token if it was on its own line
+  // 'after' is the hypothetical indentation of a comment after that token, and of the next token if it does not have a set indentation
+  private void SetIndentations(IToken token, int before = -1, int sameLineBefore = -1, int after = -1) {
+    if (token is IncludeToken || token.line == 0 && token.col == 0) { // Just ignore this token.
+      return;
+    }
+    if (before >= 0) {
+      posToIndentBefore[token.pos] = before;
+    }
+
+    if (sameLineBefore >= 0) {
+      posToIndentLineBefore[token.pos] = sameLineBefore;
+    }
+
+    if (after >= 0) {
+      posToIndentAfter[token.pos] = after;
+    }
+  }
+
+  // functions, methods, predicates, iterators...
+  void SetMethodLikeIndent(IToken startToken, IEnumerable<IToken> ownedTokens, int indent) {
```

Comment:
Hey, I do find this quite hard to read. It doesn't give me an overview of what the end result looks like by reading the code. Syntactically it reminds me of a hand written parser that switches on symbols to update a state, as opposed to a parser defined using a grammar.

Did you happen to read this paper about a formatter for R? http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/44667.pdf

It uses combinators to define the formatter, and includes a choice combinator to enable allowing formatting in different ways, which allows things like changing the formatting depending on whether there's a newline after `<` or not, like you're doing here. This paper is more aggressive in its formatting though since it also inserts newlines, but you could replace the search algorithm that determines when to insert newlines by one that uses the existing newlines to decide which direction to go in.

Reply:
> Hey, I do find this method hard to read. It doesn't give me an overview of what the end result looks like by reading the code. It reminds me of a hand written parser that switches on symbols to update a state, as opposed to a parser defined using a grammar.

Please refer to the tests. I added a comment in front of this method that say to refer to this test:
https://github.com/dafny-lang/dafny/blob/42ba60dd232fddd82b27df4afbaa6fd8c7bfc4e6/Source/DafnyPipeline.Test/Formatter.cs#L123

For the rest of your comment, I would have loved to do something much more formal and functional rather than low-level code, like in the papers you describe.
But first, I did not want to spend few years like for scalafmt's thesis, I wanted to move as quickly as possible while keeping configurability. It occurred to me that Dafny was not designed to make the intermediate representation as simple as possible, like in Scala. It's way more complex on the front-end, because it was designed with verification in mind.
Moreover, the Dafny AST maps not very well to the Dafny source code, contrary to other languages - so many low-level optimizations in the parser, and then in the resolver - I considered this token formatter as the best value*configurability/(dev time*cost) value, and I still consider it like that.
This is not to say that an alternative approach won't be possible in the future, but that would certainly take more time. I hope that my edits to the AST and INodes might enable such opportunities in the future !

Reply:
Thanks for the explanation, makes sense.

I do want to say though that any significant code contributions involves multiple contributors, both through the code review process and in the code maintenance that follows. I think any large contribution where there are design choices warrants design discussion, and I would have preferred it if this work had involved more of that.